util/stop: a better async task interface

andreimatei opened the pull request:
See individual commits. They all improve the stopper.RunAsyncTask() interface. The only behavior change is that the DistSender's partial batches are not included in the parent's trace.

andreimatei committed:
util/stop: simplify RunLimitedAsyncTask

Implement RunLimitedAsyncTask in terms of RunAsyncTask. This also
unifies the behavior w.r.t the provided taskName - one of the two
methods was prepending an "[async]" and the other wasn't.

Release note: None

andreimatei committed:
util/stop: merge RunAsyncTask and RunLimitedAsyncTask

Add a new stoppper.RunAsyncTaskEx(), which takes an optional semaphore
argument. The new method uses an extensible set of options, as more
options are sure to come in the future.

Release note: None

andreimatei committed:
stop, kvclient: include DistSender partial batches in traces

Before this patch, since fairly recently, the spans created for
Stopper.RunAsyncTask[Ex]() were not included in the recording of the
caller's span because these spans were created with the ForkSpan(). This
patch adds an option to RunAsyncEx to make the task's span a child of
the caller's, and thus to be included in the caller's recording. This
option is used by the DistSender when sending partial batch requests,
therefore re-including these requests in higher-level traces (as they
used to be).

The reason why async tasks were detached from the caller's span
in #59815 was because of concerns about parent and children spans with
very different lifetimes becoming tied through a trace, delaying the
garbage collection of the whole trace until the last span in it
finishes. This patch gives the caller of RunAsyncTaskEx control over
this behavior; in the particular case of the DistSender, the traces are
not too long lived and the DistSender waits for the async tasks it
spawns to finish. In such circumstances, tying the allocations of the
child spans to the whole trace should be fine.

Release note (general change): A recent release removed parts of some
queries from the debugging traces of those queries. This information
(i.e. the execution of some low-level RPCs) has been re-included in the
traces.

andreimatei committed:
contextutil: add utility for ctx that doesn't inherit cancellation

Add contextutil.WithoutCancel(ctx), which returns a ctx identical to the
argument but without inheriting its cancellation signal. The patch uses
this in one caller, which results in simpler code.

Release note: None

andreimatei committed:
util/stop: add options to RunAsyncTaskEx for ctx cancelation

Add an option to not inherit the context cancelation from the caller.
Before this option, the callers had to do this themselves. It's a
frequent enough need that it makes sense for the library to offer it, in
particular in conjunction with the 2nd option below.
Also add an option to cancel the task's context on stopper quiescence.
The caller could do this itself, but again it's quite common to want
this. The particular combination of wanting to not inherit the parent's
cancelation but to want cancelation on quiescence requires care when
done by the caller, because the order of the required context
manipulation matters: the stopper.WithCancelOnQuiescence() cannot be
used outside of the task's closure if the DontInheritCancellation option
is used.
One kv caller is switched to the new interface, resulting in simpler
code.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/66387)
<!-- Reviewable:end -->

erikgrinaker reviewed (commented):
This is great, thanks for cleaning this up!

Another pattern I find myself repeating all over the place, and which gets unnecessarily verbose, is to run the task with a timeout (i.e. in combination with `contextutil.RunWithTimeout`). See e.g.:

https://github.com/cockroachdb/cockroach/blob/4e4f31a0ed1a8ea985b9ab6f72e29266b259900e/pkg/kv/txn.go#L790-L799

It would be great to have a `Timeout` parameter as well. This might get a bit more involved, since we'd have to add in some sort of error handling as well (although in the typical case we'd just want to log the error). Would be happy to submit a follow-up PR for this if you'd rather not deal with it now.

> Reviewed 1 of 1 files at r1, 9 of 9 files at r2, 4 of 4 files at r3, 3 of 3 files at r4, 4 of 4 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66387#-:-MbzgO4P7zffIkILCORG:b-8pwq6u)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei, @irfansharif, @nvanbenschoten, and @tbg)

___
*[pkg/kv/txn.go, line 788 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/66387#-MbzfEek0H7ZFRpm6dAd:-MbzfEek0H7ZFRpm6dAe:b-b5zwy7) ([raw file](https://github.com/cockroachdb/cockroach/blob/cb3de48f565586f35e97b42ed7c189ab01a428ae/pkg/kv/txn.go#L788)):*
> ```Go
> 	// actual cleanup will be independent of this context.
> 	stopper := txn.db.ctx.Stopper
> 	ctx, cancel := stopper.WithCancelOnQuiesce(contextutil.WithoutCancel(ctx))
> ```

The name may be a bit misleading here. It could be easy to read this and think "well, since the stopper sets cancellation, there's no need to remove it first" -- which would be wrong, of course. Perhaps `contextutil.Disconnect()` or `DisconnectCancel()` might be clearer.


<!-- Sent from Reviewable.io -->

tbg reviewed (approved):
Thanks for working on this Andrei! Nothing super substantial in my comments, just a few suggestions.

@erikgrinaker in your snippet above, the error is discarded and so `RunWithTimeout` won't be doing anything useful (rather than actually setting up ctx cancellation). So adding a `Timeout time.Duration` parameter should provide parity.

I think what the snippet probably wants to do is to do the error handling on the error *returned* from `RunWithTimeout`. And then once it did that, yeah, for parity the error would have to be passed through the stopper somehow (so that it could do the error decoration). But I think this is getting a little awkward, and I don't want to bake logging into the stopper. I'd stay away from this at least for this PR, happy to help out to get something done in a follow-up.

> Reviewed 1 of 1 files at r1, 9 of 9 files at r2, 4 of 4 files at r3, 3 of 3 files at r4, 4 of 4 files at r5.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66387#-:-Mc92EeS48crl30F6u5f:b-aev3mz)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei, @erikgrinaker, @irfansharif, and @nvanbenschoten)

___
*[pkg/kv/txn.go, line 788 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/66387#-MbzfEek0H7ZFRpm6dAd:-Mc9-3z50al3vyIYnYOo:bde6wd6) ([raw file](https://github.com/cockroachdb/cockroach/blob/cb3de48f565586f35e97b42ed7c189ab01a428ae/pkg/kv/txn.go#L788)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

The name may be a bit misleading here. It could be easy to read this and think "well, since the stopper sets cancellation, there's no need to remove it first" -- which would be wrong, of course. Perhaps `contextutil.Disconnect()` or `DisconnectCancel()` might be clearer.
</blockquote></details>

Hmm, I'm not sure. Cancellation (like all standard ops on `Context`) is additive. As you know

```go
ctx, cancel := context.WithDeadline(context.Background(), 1*time.Second)
ctx, cancel = context.WithDeadline(ctx, 5*time.Second)
```

will result in `ctx` having a `1s` deadline. Wouldn't your argument also imply that this is unexpected? "Oh, I'm setting a 5s deadline, which overrides the 1s deadline, so I can just skip that step"?
___
*[pkg/kv/txn.go, line 792 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/66387#-Mc912EN0oJNMn7l0hpZ:-Mc912EN0oJNMn7l0hp_:bv0wqfc) ([raw file](https://github.com/cockroachdb/cockroach/blob/99f1229a4971662e8deaf656e0b36368f54c9e63/pkg/kv/txn.go#L792)):*
> ```Go
> 		stop.TaskOpts{
> 			TaskName:                "aync-rollback",
> 			DontInheritCancellation: true,
> ```

Forgetting about existing code that would change behavior, what would we want the defaults for `DontInheritCancellation` and `CancelOnQuiesce` to be? When we originally introduced the `Stopper` the idea was that it would be the main workhorse in a graceful shutdown. That hasn't hold true - I think when we quiesce it we mostly want "all tasks to go away promptly", which suggests that the default for `CancelOnQuiesce` should be true. For `DontInheritCancellation`, it might be the opposite? If you're launching an async task you usually have to expect that it will want to outlive the caller's context cancellation, and in fact we have had bugs in this area (in intent resolution I think).

While we're here, should we make that change? I'm suggesting we make the change and make sure all callers are updated to keep their old behavior.
___
*[pkg/util/contextutil/uncancelled_context.go, line 42 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/66387#-Mc903-MCx4iVsepV9ju:-Mc903-MCx4iVsepV9jv:b-8dd8rn) ([raw file](https://github.com/cockroachdb/cockroach/blob/cb3de48f565586f35e97b42ed7c189ab01a428ae/pkg/util/contextutil/uncancelled_context.go#L42)):*
> ```Go
> 
> // WithoutCancel returns a context that doesn't inherit the cancellation of its
> // parent, and so it can never be canceled.
> ```

If you eat a CI cycle anyway, maybe add that the values are inherited which is the whole point of this method.
___
*[pkg/util/stop/stopper.go, line 401 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/66387#-Mc8wkDIDcUb8oaw925a:-Mc8wkDIDcUb8oaw925b:b6d6zv3) ([raw file](https://github.com/cockroachdb/cockroach/blob/37cabca9514a9aed74bce45d2a8ebf4a1b8b9e42/pkg/util/stop/stopper.go#L401)):*
> ```Go
> 	}
> 
> 	taskName := asyncTaskNamePrefix + opt.TaskName
> ```

Is `asyncTaskNamePrefix` worth it? It's extra allocations and also means that the final task name will not be that put into the span, which can be mildly confusing. We could take this refactor as the opportunity to phase this out, seeing how you point out that it was previously inconsistently applied.
___
*[pkg/util/stop/stopper.go, line 352 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/66387#-Mc8xEBk5EE5lc9zfrYE:-Mc8xEBk5EE5lc9zfrYF:bk2z6t7) ([raw file](https://github.com/cockroachdb/cockroach/blob/56bcc5672344f00039088453996f1c9138683031/pkg/util/stop/stopper.go#L352)):*
obligatory nit: even with a FollowsFrom relationship, it's the same trace. You could call this `ChildSpan bool` to avoid this.
In the comment, we should also be more precise (this stuff is intricate enough as is). Whether the relationship is `ChildOf` or `FollowsFrom` really doesn't matter. What we're after is to use `WithParentAndAutoCollection` instead of `WithParentAndManualCollection`, because the former will reference the span created for the async operation from the parent span and thus makes it retrievable through its recording. So I'd say something like

> ChildSpan, if set, creates the tracing span for this task via `tracing.ChildSpan` instead of `tracing.ForkSpan`. This makes the task's span available via the parent span's recording (it is created with the `WithParentAndAutoCollection` option instead of `WithParentAndManualCollection`). It also leads to a `ChildOf` relationship instead of a `FollowsFrom` relationship to be used for the task's span, which typically implies a non-binding expectation that the parent span will outlive the task's span, i.e. that the parent will wait for the task to complete.

> Regardless ....

> Setting ...
___
*[pkg/util/stop/stopper.go, line 360 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/66387#-Mc90jVK4Jtek6CaV3sb:-Mc90jVK4Jtek6CaV3sc:b-kan5fu) ([raw file](https://github.com/cockroachdb/cockroach/blob/99f1229a4971662e8deaf656e0b36368f54c9e63/pkg/util/stop/stopper.go#L360)):*
> ```Go
> 	DontInheritCancellation bool
> 	// CancelOnQuiesce runs the task in a context that gets canceled when the
> 	// stopper quiesces. If DontInheritCancellation is not set, then the tasks
> ```

task's
___
*[pkg/util/stop/stopper.go, line 361 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/66387#-Mc90obh6mRjJJMIl_g5:-Mc90obh6mRjJJMIl_g6:bacegeg) ([raw file](https://github.com/cockroachdb/cockroach/blob/99f1229a4971662e8deaf656e0b36368f54c9e63/pkg/util/stop/stopper.go#L361)):*
> ```Go
> 	// CancelOnQuiesce runs the task in a context that gets canceled when the
> 	// stopper quiesces. If DontInheritCancellation is not set, then the tasks
> 	// context inherits the parent's cancelation, as well as the stopper's
> ```

cancellation


<!-- Sent from Reviewable.io -->

erikgrinaker reviewed (commented):
> I think what the snippet probably wants to do is to do the error handling on the error returned from RunWithTimeout. And then once it did that, yeah, for parity the error would have to be passed through the stopper somehow (so that it could do the error decoration). But I think this is getting a little awkward, and I don't want to bake logging into the stopper. I'd stay away from this at least for this PR, happy to help out to get something done in a follow-up.

I think a better approach for `RunWithTimeout` generally is to add e.g. `contextutil.WithTimeout()` which returns a context that decorates the error on `Err()`. Then we wouldn't have to use a closure, and we wouldn't have to do any error handling in `RunAsyncTaskEx()`.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66387#-:-Mc9Esvh1CzKEthdLSQD:blxetso)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei, @irfansharif, @nvanbenschoten, and @tbg)

___
*[pkg/kv/txn.go, line 788 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/66387#-MbzfEek0H7ZFRpm6dAd:-Mc9FOYXEG0IXZKtR7dg:b-yauynf) ([raw file](https://github.com/cockroachdb/cockroach/blob/cb3de48f565586f35e97b42ed7c189ab01a428ae/pkg/kv/txn.go#L788)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Hmm, I'm not sure. Cancellation (like all standard ops on `Context`) is additive. As you know

```go
ctx, cancel := context.WithDeadline(context.Background(), 1*time.Second)
ctx, cancel = context.WithDeadline(ctx, 5*time.Second)
```

will result in `ctx` having a `1s` deadline. Wouldn't your argument also imply that this is unexpected? "Oh, I'm setting a 5s deadline, which overrides the 1s deadline, so I can just skip that step"?
</blockquote></details>

I think it's just the "without cancel/with cancel" wording in that example that looks a bit odd. Not a big deal though, just a suggestion.


<!-- Sent from Reviewable.io -->

tbg reviewed (commented):
Yeah, unfortunately due to the history of `context.Context` it's a bit unclear if something will break. `Err()` is really only supposed to return `context.DeadlineExceeded` or `context.Canceled`; it says so right in the comment. Since the errors will also pop up in our vendored code, this isn't safe without a giant audit (and re-audit whenever deps change...)

What we could do is to use a convention:

```go
func (ctx *crdbCtx) setErr(err error) {
  if errors.Is(err, context.DeadlineExceeded) {
    ctx.err = context.DeadlineExceeded
  } else {
    ctx.err = context.Canceled
  }
  ctx.origErr = err
}

func (ctx *crdbCtx) Value(k interface{}) interface{} {
  if k == ExtErrKey{} && ctx.origErr != nil {
    return ctx.origErr
  }
  return ctx.wrapped.Value(k)
}
```

and you could do this:

```go
ctx, _ := contextutil.WithHelpfulTimeout(context.Background, 1*time.Second)
time.Sleep(2*time.Second)
require.Equal(t, context.DeadlineExceeded, ctx.Err())
require.True(t, errors.Is(contextutil.ExtendedError(ctx), (*contextutil.TimeoutError)(nil)))
```

This isn't that closely related to the stopper any more, but it's a real improvement over `RunWithTimeout` in my opinion; you don't need a closure which is great in the common case. We "just" need to get in the habit of using `ExtendedError` instead of` ctx.Err()`, but hey we have linted that sort of thing to death more than once.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66387#-:-Mc9QtX18wLYRkukWSk2:b-7mbruk)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei, @irfansharif, @nvanbenschoten, and @tbg)



<!-- Sent from Reviewable.io -->

andreimatei committed:
stop,contextutil: optimize Stopper.WithCancelOnQuiesce

This patch makes WithCancelOnQuiesce be a no-op when called on a ctx
that was already inheriting cancellation on the quiescing of the
respective Stopper, avoiding multiple allocations. This optimization is
done with an eye towards making cancellation on quiescence the default
for stopper tasks.

Release note: None

andreimatei reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66387)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @erikgrinaker, @irfansharif, @nvanbenschoten, and @tbg)

___
*[pkg/kv/txn.go, line 788 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/66387#-MbzfEek0H7ZFRpm6dAd:-McAQqvqBj0doEkD_im2:bi2eb0f) ([raw file](https://github.com/cockroachdb/cockroach/blob/cb3de48f565586f35e97b42ed7c189ab01a428ae/pkg/kv/txn.go#L788)):*
> I think it's just the "without cancel/with cancel" wording in that example that looks a bit odd. Not a big deal though, just a suggestion.

The odd alliteration goes away in the next commit. 
I've considered various names for `WithoutCancel`, and it still seems to me like most suggestive option.
___
*[pkg/kv/txn.go, line 792 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/66387#-Mc912EN0oJNMn7l0hpZ:-McASQX-B-Zcb9BnZARH:bd51gyq) ([raw file](https://github.com/cockroachdb/cockroach/blob/99f1229a4971662e8deaf656e0b36368f54c9e63/pkg/kv/txn.go#L792)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Forgetting about existing code that would change behavior, what would we want the defaults for `DontInheritCancellation` and `CancelOnQuiesce` to be? When we originally introduced the `Stopper` the idea was that it would be the main workhorse in a graceful shutdown. That hasn't hold true - I think when we quiesce it we mostly want "all tasks to go away promptly", which suggests that the default for `CancelOnQuiesce` should be true. For `DontInheritCancellation`, it might be the opposite? If you're launching an async task you usually have to expect that it will want to outlive the caller's context cancellation, and in fact we have had bugs in this area (in intent resolution I think).

While we're here, should we make that change? I'm suggesting we make the change and make sure all callers are updated to keep their old behavior.
</blockquote></details>

I too think that `CancelOnQuiesce` should be the default. But I've had performance concerns - particularly because `CancelOnQuiesce` is currently wasteful when called on a ctx that itself inherits from a `CancelOnQuiesce` one - which would become very common. I've appended a commit that improves that implementation. But it gets into a more hands-on approach about the implementations of `context.Context` that we use, given that the stdlib one is not great. I think you'll welcome this, because you've been talking above about why that standard cancellation facilities suck. So please PTAL at that commit, and then I'll invert the defaults.

I'm also thinking about tying the default of `DontInheritCancellation` to the value of `ChildSpan`. Do you have thoughts on that? I'm thinking perhaps adding a simple way of creating a `TaskOpts` that sets both `ChildSpan` and `InheritCancelation` - perhaps as a `TaskOpts` ctor with a suggestive name (`SyncChildTask()`?).
___
*[pkg/util/stop/stopper.go, line 401 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/66387#-Mc8wkDIDcUb8oaw925a:-McAKcYjDohfLzLbKNLG:bk0f2k9) ([raw file](https://github.com/cockroachdb/cockroach/blob/37cabca9514a9aed74bce45d2a8ebf4a1b8b9e42/pkg/util/stop/stopper.go#L401)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Is `asyncTaskNamePrefix` worth it? It's extra allocations and also means that the final task name will not be that put into the span, which can be mildly confusing. We could take this refactor as the opportunity to phase this out, seeing how you point out that it was previously inconsistently applied.
</blockquote></details>

Indeed it allocates. I've removed it.
___
*[pkg/util/stop/stopper.go, line 352 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/66387#-Mc8xEBk5EE5lc9zfrYE:-McAHWom5TGO3cbWiP_r:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/56bcc5672344f00039088453996f1c9138683031/pkg/util/stop/stopper.go#L352)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

obligatory nit: even with a FollowsFrom relationship, it's the same trace. You could call this `ChildSpan bool` to avoid this.
In the comment, we should also be more precise (this stuff is intricate enough as is). Whether the relationship is `ChildOf` or `FollowsFrom` really doesn't matter. What we're after is to use `WithParentAndAutoCollection` instead of `WithParentAndManualCollection`, because the former will reference the span created for the async operation from the parent span and thus makes it retrievable through its recording. So I'd say something like

> ChildSpan, if set, creates the tracing span for this task via `tracing.ChildSpan` instead of `tracing.ForkSpan`. This makes the task's span available via the parent span's recording (it is created with the `WithParentAndAutoCollection` option instead of `WithParentAndManualCollection`). It also leads to a `ChildOf` relationship instead of a `FollowsFrom` relationship to be used for the task's span, which typically implies a non-binding expectation that the parent span will outlive the task's span, i.e. that the parent will wait for the task to complete.

> Regardless ....

> Setting ...
</blockquote></details>

done
___
*[pkg/util/stop/stopper.go, line 360 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/66387#-Mc90jVK4Jtek6CaV3sb:-McANslc2nwh9GoBDi6s:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/99f1229a4971662e8deaf656e0b36368f54c9e63/pkg/util/stop/stopper.go#L360)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

task's
</blockquote></details>

done
___
*[pkg/util/stop/stopper.go, line 361 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/66387#-Mc90obh6mRjJJMIl_g5:-McANwpSBKu0CsW31J2l:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/99f1229a4971662e8deaf656e0b36368f54c9e63/pkg/util/stop/stopper.go#L361)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

cancellation
</blockquote></details>

done
___
*[pkg/util/contextutil/uncancelled_context.go, line 42 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/66387#-Mc903-MCx4iVsepV9ju:-McANWiA1Wz8DObr4GEk:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/cb3de48f565586f35e97b42ed7c189ab01a428ae/pkg/util/contextutil/uncancelled_context.go#L42)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

If you eat a CI cycle anyway, maybe add that the values are inherited which is the whole point of this method.
</blockquote></details>

done


<!-- Sent from Reviewable.io -->

andreimatei reviewed (commented):
I haven't done anything about deadlines in this PR, and I might leave the topic to another PR.
But I've appended another PR with some context hacking - PTAL. I sympathize with the Tobi's sentiment that the stdlib's context.Context is not a great library. I'd like to take steps towards being more aggressive in overcoming it's limitations around cancellation.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66387#-:-McBZa1n0hXFe0VPd-YH:b-y3pyf5)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @erikgrinaker, @irfansharif, @nvanbenschoten, and @tbg)



<!-- Sent from Reviewable.io -->

tbg reviewed (commented):
The broad strokes of this look good but unfortunately I ran out of time to give `crdbctx.go` a close look (which I think it deserves). LGTM mod that file, perhaps @erikgrinaker can review that.

> Reviewed 13 of 13 files at r6, 10 of 10 files at r7, 6 of 6 files at r8, 3 of 3 files at r9, 4 of 4 files at r10, 4 of 4 files at r11.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66387#-:-McPi1r67ShbUymu7_3y:b-3chpjv)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei, @erikgrinaker, @irfansharif, @nvanbenschoten, and @tbg)

___
*[pkg/util/contextutil/crdbctx.go, line 77 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/66387#-McPhtjK47k8kSNBlhNT:-McPhtjK47k8kSNBlhNU:b-gqrpis) ([raw file](https://github.com/cockroachdb/cockroach/blob/a4aa979204d4f34cb3d1a9ea09a632a0c3cf7e70/pkg/util/contextutil/crdbctx.go#L77)):*
> ```Go
> 	uncParent := getCancellationBarrierParent(ctx)
> 	if uncParent == nil {
> 		return true
> ```

Comment here would be good, this is unintuitive at first.
___
*[pkg/util/contextutil/crdbctx.go, line 168 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/66387#-McPgW-s6j0g08caw5XZ:-McPgW-s6j0g08caw5X_:b3etzob) ([raw file](https://github.com/cockroachdb/cockroach/blob/a4aa979204d4f34cb3d1a9ea09a632a0c3cf7e70/pkg/util/contextutil/crdbctx.go#L168)):*
> ```Go
> var cancellationBarrierKey = new(int)
> 
> // getUncanceledParent returns the innermost uncanceledContext amongst ctx's
> ```

there are a few references to uncanceledContext, is this detritus?
___
*[pkg/util/stop/stopper.go, line 352 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/66387#-McPfpqO7RxJrUkKug3k:-McPfpqO7RxJrUkKug3l:b32v2d) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7fb43958bb0025b14136e2d052297d1167c1fc6/pkg/util/stop/stopper.go#L352)):*
> ```Go
> 	// ChildSpan, if set, creates the tracing span for the task via
> 	// tracing.ChildSpan() instead of tracing.ForkSpan. This makes the task's span
> 	// be part of the parent span's recording ((it is created with the
> ```

((


<!-- Sent from Reviewable.io -->

erikgrinaker reviewed (commented):
First quick pass over `crdbctx.go`, dumping some preliminary comments.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66387#-:-McPuon3ABScWQTN3NLG:b4b9rar)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei, @irfansharif, @nvanbenschoten, and @tbg)

___
*[pkg/util/contextutil/crdbctx.go, line 62 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/66387#-McPqsvOCbTlQCmmpb_4:-McPqsvOCbTlQCmmpb_5:b-terll3) ([raw file](https://github.com/cockroachdb/cockroach/blob/a4aa979204d4f34cb3d1a9ea09a632a0c3cf7e70/pkg/util/contextutil/crdbctx.go#L62)):*
> ```Go
> // certain events (e.g. stopper quiescence) and then, in child tasks, checking
> // whether we're still running inside such a context.
> func Tagged(ctx context.Context, key interface{}, opt CtxWalkOption) bool {
> ```

Is `opt` important enough that it warrants being required for every call, or could we add an additional method for it when needed?

Also, please use `tag` instead of `key`, for symmetry with `WithTag()` and to make the code easier to follow.
___
*[pkg/util/contextutil/crdbctx.go, line 63 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/66387#-McPuY8u2AyOIN2b3wJR:-McPuY8vB5anHABot9kk:b-6qu4kp) ([raw file](https://github.com/cockroachdb/cockroach/blob/a4aa979204d4f34cb3d1a9ea09a632a0c3cf7e70/pkg/util/contextutil/crdbctx.go#L63)):*
> ```Go
> // whether we're still running inside such a context.
> func Tagged(ctx context.Context, key interface{}, opt CtxWalkOption) bool {
> 	p, ok := ctx.Value(key).(*crdbCtx)
> ```

nit: could call this e.g. `tagged` or something, for clarity.
___
*[pkg/util/contextutil/crdbctx.go, line 71 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/66387#-McPqG_j7R-107gG6Hcp:-McPqG_j7R-107gG6Hcq:bnwr47v) ([raw file](https://github.com/cockroachdb/cockroach/blob/a4aa979204d4f34cb3d1a9ea09a632a0c3cf7e70/pkg/util/contextutil/crdbctx.go#L71)):*
> ```Go
> 	}
> 
> 	if opt != CancellationParents {
> ```

nit: could use a `switch` here.
___
*[pkg/util/contextutil/crdbctx.go, line 126 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/66387#-McPsYl9BN3mgGDCHiL1:-McPsYl9BN3mgGDCHiL2:b-n79fa7) ([raw file](https://github.com/cockroachdb/cockroach/blob/a4aa979204d4f34cb3d1a9ea09a632a0c3cf7e70/pkg/util/contextutil/crdbctx.go#L126)):*
> ```Go
> // Value is part of the Context interface.
> func (c *crdbCtx) Value(key interface{}) interface{} {
> 	if key == c.tag {
> ```

I'm finding it very unintuitive that we're overloading the semantics of `Value` to involve tag and cancellation barriers, and it makes Tagged() harder to reason about. I suppose it may be necessary in order to integrate with the stdlib context (is it?), but it should come with a big fat comment in that case.


<!-- Sent from Reviewable.io -->

tbg reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66387)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei, @erikgrinaker, @irfansharif, @nvanbenschoten, and @tbg)

___
*[pkg/util/contextutil/crdbctx.go, line 20 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/66387#-MciLmxPBVIBOn2vkTiA:-MciLmxPBVIBOn2vkTiB:byzllee) ([raw file](https://github.com/cockroachdb/cockroach/blob/a4aa979204d4f34cb3d1a9ea09a632a0c3cf7e70/pkg/util/contextutil/crdbctx.go#L20)):*
> ```Go
> 
> // WithTag returns a context tagged with the provided tag. Calling Tagged(tag)
> // on the returned context or on a child context will return true.
> ```

This needs to explain how

```go
ctx  = WithTag(ctx, something)
require.True(t, Tagged(ctx))
```

is different from what you can do with vanilla `Context`:

```go
ctx = context.WithValue(ctx, something, something)
_, ok := context.Value(something).(*somethingType)
require.True(t, ok)
```

The answer (only one I can come up with) is that you really care about the `CancellationParents` option. But you can get that with vanilla too:

```go
ctx = context.WithValue(ctx, something, something)

ctx = context.WithoutCancel(context.WithValue(something, nil))
require.Nil(t, context.Value(something))
```
___
*[pkg/util/contextutil/crdbctx.go, line 42 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/66387#-MciMnkb2LNeK7b8i6YT:-MciMnkb2LNeK7b8i6YU:b-wj5p3i) ([raw file](https://github.com/cockroachdb/cockroach/blob/a4aa979204d4f34cb3d1a9ea09a632a0c3cf7e70/pkg/util/contextutil/crdbctx.go#L42)):*
> ```Go
> type CtxWalkOption int
> 
> const (
> ```

This stuff just feels overly complex. Let's go back to the problem we actually want to solve. I'm sure we can do something simpler.
___
*[pkg/util/contextutil/crdbctx.go, line 126 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/66387#-McPsYl9BN3mgGDCHiL1:-MciOb5gBjTU9rnAeNo_:b32v57) ([raw file](https://github.com/cockroachdb/cockroach/blob/a4aa979204d4f34cb3d1a9ea09a632a0c3cf7e70/pkg/util/contextutil/crdbctx.go#L126)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

I'm finding it very unintuitive that we're overloading the semantics of `Value` to involve tag and cancellation barriers, and it makes Tagged() harder to reason about. I suppose it may be necessary in order to integrate with the stdlib context (is it?), but it should come with a big fat comment in that case.
</blockquote></details>

+1


<!-- Sent from Reviewable.io -->

andreimatei reviewed (commented):
Tobi, I have received an email with your last round of comments but, for whatever reason, they don't show up in Reviewable. In any case, they were all around WithTag/Tagged. I've removed those in favor of a simpler `InheritsCancellation` function that can be implemented only through the custom `uncanceledContext`, without the need for another custom "tagged" context. PTAL.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66387#-:-McjoAGR2L1NrZjJjnGl:b-7u3ajo)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @erikgrinaker, @irfansharif, @nvanbenschoten, and @tbg)

___
*[pkg/util/stop/stopper.go, line 352 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/66387#-McPfpqO7RxJrUkKug3k:-Mcjnz8TCfaZnUPiik-F:b3ksf) ([raw file](https://github.com/cockroachdb/cockroach/blob/f7fb43958bb0025b14136e2d052297d1167c1fc6/pkg/util/stop/stopper.go#L352)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

((
</blockquote></details>

d
___
*[pkg/util/contextutil/crdbctx.go, line 62 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/66387#-McPqsvOCbTlQCmmpb_4:-McjD8wGBfgOnQkXZHXJ:b-hfpv0l) ([raw file](https://github.com/cockroachdb/cockroach/blob/a4aa979204d4f34cb3d1a9ea09a632a0c3cf7e70/pkg/util/contextutil/crdbctx.go#L62)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

Is `opt` important enough that it warrants being required for every call, or could we add an additional method for it when needed?

Also, please use `tag` instead of `key`, for symmetry with `WithTag()` and to make the code easier to follow.
</blockquote></details>

removed
___
*[pkg/util/contextutil/crdbctx.go, line 63 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/66387#-McPuY8u2AyOIN2b3wJR:-McjDm-A7RF0PBjFlKkm:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/a4aa979204d4f34cb3d1a9ea09a632a0c3cf7e70/pkg/util/contextutil/crdbctx.go#L63)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

nit: could call this e.g. `tagged` or something, for clarity.
</blockquote></details>

done
___
*[pkg/util/contextutil/crdbctx.go, line 71 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/66387#-McPqG_j7R-107gG6Hcp:-McjE-Uz403iM29yTpxq:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/a4aa979204d4f34cb3d1a9ea09a632a0c3cf7e70/pkg/util/contextutil/crdbctx.go#L71)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

nit: could use a `switch` here.
</blockquote></details>

done
___
*[pkg/util/contextutil/crdbctx.go, line 77 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/66387#-McPhtjK47k8kSNBlhNT:-McjEQZP2SlsA8-8fGIg:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/a4aa979204d4f34cb3d1a9ea09a632a0c3cf7e70/pkg/util/contextutil/crdbctx.go#L77)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Comment here would be good, this is unintuitive at first.
</blockquote></details>

done
___
*[pkg/util/contextutil/crdbctx.go, line 126 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/66387#-McPsYl9BN3mgGDCHiL1:-McjnQ3G1dqV0zwZTw08:bghzm2) ([raw file](https://github.com/cockroachdb/cockroach/blob/a4aa979204d4f34cb3d1a9ea09a632a0c3cf7e70/pkg/util/contextutil/crdbctx.go#L126)):*
<details><summary><i>Previously, erikgrinaker (Erik Grinaker) wrote…</i></summary><blockquote>

I'm finding it very unintuitive that we're overloading the semantics of `Value` to involve tag and cancellation barriers, and it makes Tagged() harder to reason about. I suppose it may be necessary in order to integrate with the stdlib context (is it?), but it should come with a big fat comment in that case.
</blockquote></details>

Added a comment. FWIW, the stdlib does the same kind of trick for its context implementations.
___
*[pkg/util/contextutil/crdbctx.go, line 168 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/66387#-McPgW-s6j0g08caw5XZ:-McjngrA0bAPQgDFwfOw:b2ni76j) ([raw file](https://github.com/cockroachdb/cockroach/blob/a4aa979204d4f34cb3d1a9ea09a632a0c3cf7e70/pkg/util/contextutil/crdbctx.go#L168)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

there are a few references to uncanceledContext, is this detritus?
</blockquote></details>

n/a


<!-- Sent from Reviewable.io -->

tbg reviewed (commented):
For posterity, the email was from a Github review, see https://github.com/cockroachdb/cockroach/pull/66387#pullrequestreview-688331484, somehow they're also in reviewable? Not sure what's going on but hitting Publish now & then reviewing your changes.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66387#-:-Mck2_My-WW8mkVmMFXC:bf76v1x)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei, @erikgrinaker, @irfansharif, @nvanbenschoten, and @tbg)



<!-- Sent from Reviewable.io -->

andreimatei reviewed (commented):
@tbg, we've discussed offline yesterday replacing the `uncanceledContext` guy with a facility to copy parts of a context into a new one, and thus get rid of the parent's cancellation that way (by essentially creating a new context). Thinking about it again, I'm no longer sure what we'd gain; I wonder if we fooled ourselves. 
The problem with the `uncanceledContext` is that the `InheritsCancellation` function only works if `uncanceledContext` is the only context implementation playing tricks with `Done()/Error()`. But the thing is, the implementation of `InheritsCancellation` based on the ctx copy scheme would have exactly the same problem. To recap, the point of the last commit is an optimization: making `Stopper.WithCancelOnQuiesce` a no-op (in particular, no allocations) if the parent already has a `WithCancelOnQuiesce` on the chain. Assuming that we want this optimization, we need a way to figure out whether a particular cancellation is inherited. How would we do that in the ctx-copy world? Presumably we'd simply look to see if the child ctx is annotated with the `Stopper` and if it is, assume that there's nobody on the chain that overrode `Done()/Error()`. So, we've gotten to the same assumption as the one `InheritsCancellation()` does in the current patch. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66387#-:-MdN2toaEHrQw8qvEZWX:b-xwn048)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei, @erikgrinaker, @irfansharif, @nvanbenschoten, and @tbg)



<!-- Sent from Reviewable.io -->

tbg reviewed (commented):
Certainly possible that we're fooling ourselves. My reasoning is that we've agreed that it's invalid to "uncancel" a context and so rely on this being true (i.e. part of implementing the interface, in some unspoken sense). If someone else does it, we're screwed anyway; the whole point of contexts is that cancellation is inherited. "Uncancelling" really means, start anew from `context.Background`.

> Reviewed 7 of 10 files at r13, 4 of 7 files at r14, 1 of 5 files at r16, 7 of 7 files at r17.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66387#-:-MdTNHX48C4-Ym8YC9BJ:b9zll6n)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @erikgrinaker, @irfansharif, @nvanbenschoten, and @tbg)



<!-- Sent from Reviewable.io -->
